pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -0,0 +1,376 @@
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnetset", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnetset", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnetset", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// createSubnet creates subnet managed by subnetset
+func (r *SubnetSetReconciler) createSubnet(subnetset *v1alpha1.SubnetSet, name string) (*v1alpha1.Subnet, error) {
+	var subnet *v1alpha1.Subnet
+	if err := retry.OnError(retry.DefaultRetry, func(err error) bool {
+		return err != nil
+	}, func() error {
+		blockDeletion := true
+		obj := &v1alpha1.Subnet{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: subnetset.Namespace,
+				Name:      name,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         "v1alpha1",
+						Kind:               "subnetsets.nsx.vmware.com",
+						Name:               subnetset.Name,
+						UID:                subnetset.UID,
+						Controller:         nil,
+						BlockOwnerDeletion: &blockDeletion,
+					},
+				},
+			},
+			Spec: v1alpha1.SubnetSpec{
+				IPv4SubnetSize: subnetset.Spec.IPv4SubnetSize,
+				AccessMode:     subnetset.Spec.AccessMode,
+			},
+		}
+		if err := r.Client.Create(context.Background(), obj); err != nil {
+			return err
+		}
+		subnet = obj
+		return nil
+	}); err != nil {
+		log.Error(err, "failed to create subnet", "Namespace", subnetset.Namespace, "Name", name)
+		return nil, err
+	}
+	return subnet, nil
+}
+
+// getAvailableSubnets returns available Subnet under SubnetSet, and creates Subnet if necessary.
+func (r *SubnetSetReconciler) getAvailableSubnet(subnetSet *v1alpha1.SubnetSet) (*v1alpha1.Subnet, error) {
+	subnetList, err := r.listSubnets(subnetSet)
+	if err != nil {
+		return nil, err
+	}
+	var allocatedSubnet *v1alpha1.Subnet
+	for _, subnet := range subnetList.Items {
+		if nums, _ := r.Service.GetAvailableIPNum(&subnet); nums > 0 {
+			allocatedSubnet = &subnet
+			break
+		}
+	}
+	if allocatedSubnet == nil {
+		log.Info("no available subnet found, creating new subnet")
+		// TODO generate subnet name here
+		subnet, err := r.createSubnet(subnetSet, "subnetName")
+		if err != nil {
+			log.Error(err, "failed to create Subnet")
+			return nil, err
+		}
+		allocatedSubnet = subnet
+	}
+	return allocatedSubnet, nil
+}
+
+// listSubnets lists all subnets under subnetset
+func (r *SubnetSetReconciler) listSubnets(subnetset *v1alpha1.SubnetSet) (*v1alpha1.SubnetList, error) {
+	subnets := &v1alpha1.SubnetList{}
+	if err := retry.OnError(retry.DefaultRetry, func(err error) bool {
+		return err != nil
+	}, func() error {
+		if err := r.Client.List(context.Background(), subnets,
+			client.MatchingFields{"metadata.ownerReferences": string(subnetset.UID)}); err != nil {
```

Comment:
for SubnetSet CR, do we create Subnet CR? I thought we create NSX subnet directly.

Reply:
yes, in current implementation, subnetset controller creates a subnet CR that belongs to subnetset and delegate the creation of NSX subnet to subnet controller. subnetset controller doesn't create NSX subnet directly.

Reply:
Updated to create subnet directly.